src/rust/ide/enso-protocol/src/project_manager.rs

```diff
@@ -0,0 +1,382 @@
+//! Client library for the JSON-RPC-based Project Manager service.
+
+use crate::prelude::*;
+
+use crate::types::UTCDateTime;
+use json_rpc::api::Result;
+use json_rpc::Handler;
+use json_rpc::make_rpc_methods;
+use futures::Stream;
+use serde::Serialize;
+use serde::Deserialize;
+use std::future::Future;
+use uuid::Uuid;
+
+
+
+// =============
+// === Event ===
+// =============
+
+// Project Manager has no notifications, so we create a dummy Notification type for it.
+type Notification = ();
+
+/// Event emitted by the Project Manager `Client`.
+pub type Event = json_rpc::handler::Event<Notification>;
+
+
+
+// ===================
+// === RPC Methods ===
+// ===================
+
+make_rpc_methods! {
+/// An interface containing all the available project management operations.
+trait API {
+    /// Request the project picker to open a specified project. This operation also
+    /// includes spawning an instance of the language server open on the specified project.
+    #[MethodInput=OpenProjectInput,rpc_name="project/open",result=open_project_result,
+    set_result=set_open_project_result]
+    fn open_project(&self, project_id:Uuid) -> OpenProjectResponse;
+
+    /// Request the project picker to close a specified project. This operation
+    /// includes shutting down the language server gracefully so that it can persist state to disk
+    /// as needed.
+    #[MethodInput=CloseProjectInput,rpc_name="project/close",result=close_project_result,
+    set_result=set_close_project_result]
+    fn close_project(&self, project_id:Uuid) -> ();
+
+    /// Request the project picker to list the user's most recently opened projects.
+    #[MethodInput=ListRecentProjectsInput,rpc_name="project/listRecent",
+    result=list_recent_projects_result,set_result=set_list_recent_projects_result]
+    fn list_recent_projects(&self, number_of_projects:u32) -> ProjectListResponse;
+
+    /// Request the creation of a new project.
+    #[MethodInput=CreateProjectInput,rpc_name="project/create",result=create_project_result,
+    set_result=set_create_project_result]
+    fn create_project(&self, name:String) -> CreateProjectResponse;
+
+    /// Request the deletion of a project.
+    #[MethodInput=DeleteProjectInput,rpc_name="project/delete",result=delete_project_result,
+    set_result=set_delete_project_result]
+    fn delete_project(&self, project_id:Uuid) -> ();
+
+    /// Request a list of sample projects that are available to the user.
+    #[MethodInput=ListSamplesInput,rpc_name="project/listSample",result=list_samples_result,
+    set_result=set_list_samples_result]
+    fn list_samples(&self, num_projects:u32) -> ProjectListResponse;
+}
+}
```

Comment:
make it one ln pls. In case we are not indenting code like here, we close it everywhere else with `}}` instead.